enhancements/single-node-production-deployment-approach.md

```diff
@@ -0,0 +1,497 @@
+  will be described in a separate enhancement.
+
+## Proposal
+
+After the [capabilities API is
+introduced](https://github.com/openshift/enhancements/pull/555), all
+teams developing OpenShift components will need to consider how their
+components should be configured when deployed and used when the
+high-availability capability is disabled.
+
+In the telco RAN use case, high-availability is typically achieved by
+having multiple sites provide coverage to overlapping geographic
+areas. Therefore, use of cluster-based high-availability features will
+be limited (for example, by running a single API service).
+
+Failures in edge deployments are frequently resolved by re-imaging or
+physically replacing the entire host. Combining this fact with the
+previous observation about the approach to providing highly-available
+services lets us draw the conclusion that in-place upgrades do not
+need to be supported by the first iteration of this configuration.
+
+In a multi-node cluster, operators may safely trigger rolling reboots
+or restarts of the control plane services without taking the cluster
+offline. In a single-node deployment, any such restart would take some
+or all of the services offline. The deployment or upgrade of a cluster
+may include several host reboots, clustered together in time, until
+the operators stabilize. Users should wait for this period of
+instability before installing their workloads.
+
+The single-node configuration of OpenShift will be sufficiently
+generic to cater to a variety of edge computing use cases. As such,
+OpenShift's usual cluster configuration mechanisms will be favored
+where there is a likelihood that there will be edge computing use case
+with differing requirements.  For example, there is no expectation
+that single-node deployments will use the real-time kernel by
+default -- this will continue to be a `MachineConfig` choice as per
+[enhancements/support-for-realtime-kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md). *See open questions*
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+deployment, with some limitations (see non-goals above). The user
+will not require special support exceptions to receive technical assistance
+for the features supported by the configuration.
+
+### Implementation Details/Notes/Constraints
+
+Some OpenShift components require a minimum of 2 or 3 nodes. The new
+cluster capabilities API will be used to configure these components as
+appropriate for a single node.
+
+#### installer
+
+We will add logic to the installer to create a manifest to set up the
+singleton of the new `capabilities` API resource. The
+`highAvailabilityMode` will be set to `full` by default and `none`
+when the number of replicas for the control plane is `1`.
```

Comment:
This leaves open the possibility of single control plane node and multiple workers? In that scenario, we wouldn't want MCO to skip drain when rebooting workers?

Reply:
I've tried to update this. I'm not sure if we want MCO looking at the worker node count, and in a fully-utilized cluster there still wouldn't be anywhere to drain to and the MCO wouldn't be able to determine that on its own. So, it may still be safer to assume that in a non-HA deployment the MCO just skips the drain.

Reply:
(I think this comment was related to the next section and GitHub is now showing it in the wrong place on the review?)

Reply:
> This leaves open the possibility of single control plane node and multiple workers? 

Err...is anyone asking for that?  I don't understand the use case really for doing it.  Let's explicitly make that not supported?

Reply:
We do have one customer that is anticipating running 2 node deployments, with 1 combined control plane/worker node and 1 worker node. I wouldn't expect draining to work in that setup, either, especially if the workloads are bound to specific hardware and consuming all of the resources on the host(s). I've tried to express that in lines 240-249 of the latest draft.

Reply:
Thanks!